### PR TITLE
Added Ray class

### PR DESCRIPTION
**Pull request by Kouros26**

### CODE REVIEW

File: Core/include/PPM.h [hunk 1]
```diff
@@ -1,21 +1,18 @@
 #pragma once
 #include <fstream>
 #include <iostream>
-#include <cmath>
+#include "Vec3/Vec3.h"
 
 class PPM
```
Comment (karnkaul) [inline Core/include/PPM.h:6]: This should probably be a POD, unless you really like OOP classes.

File: Core/include/PPM.h [hunk 2]
```diff
@@ -1,21 +1,18 @@
...
 {
 public:
 
-	PPM();
-	PPM(int height, int width);
+	PPM(const int& height, const int& width);
 	~PPM();
```
Comment (karnkaul) [inline Core/include/PPM.h]: Why is a destructor needed at all?

File: Core/include/PPM.h [hunk 3]
```diff
@@ -1,21 +1,18 @@
...
 
 	struct RGB
 	{
-		unsigned char r;
-		unsigned char g;
-		unsigned char b;
+		lm::vec3 color;
```
Comment (karnkaul) [inline Core/include/PPM.h]: I'd prefer having all variables brace initialized where possible, especially in declarations.

File: Core/include/Ray.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+#ifndef RAYH
+#define RAYH
+#include "Vec3/vec3.h"
+
+class Ray
```
Comment (karnkaul) [inline Core/include/Ray.h:6]: Again, should just be a struct of data, no constructors etc needed.

File: Core/include/main.h [hunk 5]
```diff
@@ -1,4 +1,4 @@
 #pragma once
 #include "../include/PPM.h"
-#include "Vec4/Vec4.h"
-#include <iostream>
\ No newline at end of file
+#include <iostream>
+#include "../include/Ray.h"
```
Comment (karnkaul) [inline Core/include/main.h:4]: Recommend adding a proper include path and using exclusively `#include <path/to/header.hpp>` everywhere.

File: Core/src/PPM.cpp [hunk 6]
```diff
@@ -44,16 +44,16 @@ void PPM::save(const std::string& name) const
...
 				}
 		}
 
 		else
 			for (int i = 0; i < height; i++)
 				for (int j = 0; j < width; j++)
-					output.write((char*)&image[i][j], sizeof(RGB));
+					output.write(reinterpret_cast<char*>(&image[i][j]), sizeof(RGB));
 
 		output.close();
```
Comment (karnkaul) [inline Core/src/PPM.cpp]: No need to close fstreams

File: Core/src/PPM.cpp [hunk 7]
```diff
@@ -44,16 +44,16 @@ void PPM::save(const std::string& name) const
...
+					output << static_cast<int>(image[i][j].color.Y()) << ' ';
+					output << static_cast<int>(image[i][j].color.Z()) << '\n';
 				}
 		}
 
 		else
 			for (int i = 0; i < height; i++)
 				for (int j = 0; j < width; j++)
-					output.write((char*)&image[i][j], sizeof(RGB));
+					output.write(reinterpret_cast<char*>(&image[i][j]), sizeof(RGB));
```
Comment (karnkaul) [inline Core/src/PPM.cpp:51]: This may not be legal...

File: Core/include/Ray.h [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+{
+	lm::vec3 origin, direction;
+
+public:
+
+
+	Ray();
+	Ray(const lm::vec3& a, const lm::vec3& b);
+	lm::vec3 lookAt(float& t) const;
```
Comment (karnkaul) [inline Core/include/Ray.h:15]: Why does this take a `float&`?

File: Core/include/Ray.h [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+	
+	lm::vec3 getOrigin() const;
```
Comment (karnkaul) [inline Core/include/Ray.h]: Inconsistent casing: functions were snake case before

File: Core/src/PPM.cpp [hunk 10]
```diff
@@ -1,6 +1,6 @@
 #include "../include/PPM.h"
 
-PPM::PPM(int height, int width)
+PPM::PPM(const int& height, const int& width)
```
Comment (karnkaul) [inline Core/src/PPM.cpp]: No need to take types less than 64 bits by reference

File: Core/src/Ray.cpp [hunk 11]
```diff
@@ -0,0 +1,24 @@
+#include "../include/Ray.h"
+
+Ray::Ray() {}
+
+Ray::Ray(const lm::vec3& origin, const lm::vec3& direction)
+{
+	this->origin = origin;
```
Comment (karnkaul) [inline Core/src/Ray.cpp]: Rule of zero / member initializers

**Review by karnkaul — CHANGES_REQUESTED** [review]

File: Core/include/PPM.h (hunk 1 above)
Comment (Kouros26) [reply Core/include/PPM.h:6]: I do but does it affect performance ?

File: Core/include/Ray.h (hunk 8 above)
Comment (Kouros26) [reply Core/include/Ray.h:15]: Should it just be a float ?

File: Core/include/main.h (hunk 5 above)
Comment (Kouros26) [reply Core/include/main.h:4]: My CMake skills are really limited so I'm not sure how to do this using CMake only

File: Core/include/PPM.h (hunk 1 above)
Comment (karnkaul) [reply Core/include/PPM.h:6]: Not if done well. 🤓

File: Core/include/main.h (hunk 5 above)
Comment (karnkaul) [reply Core/include/main.h:4]: `target_include_directories(foo PRIVATE include src)` should be enough

File: Core/include/Ray.h (hunk 8 above)
Comment (karnkaul) [reply Core/include/Ray.h:15]: Yup

**Pull request merged**